org.eclipse.tm4e.core/src/main/java/org/eclipse/tm4e/core/model/AbstractTMModel.java

```diff
@@ -128,170 +70,203 @@ public int getSize() {
 		 * @param name
 		 *            the thread's name
 		 */
-		public BackgroundThread(String name) {
+		public TokenizerThread(String name, AbstractTMModel model) {
 			super(name);
+			this.model = model;
+			this.tokenizer = new Tokenizer(model.getGrammar());
 			setPriority(Thread.MIN_PRIORITY);
 			setDaemon(true);
 		}
 
-		/**
-		 * Suspends the caller of this method until this background thread has
-		 * emptied the dirty region queue.
-		 */
-		public void suspendCallerWhileDirty() {
-			boolean isDirty;
-			do {
-				synchronized (fDirtyRegionQueue) {
-					isDirty = fDirtyRegionQueue.getSize() > 0;
-					if (isDirty) {
+		@Override
+		public void run() {
+			if (interrupted())
+				return;
+
+			// initialize
+			int linesLength = model.getNumberOfLines();
+			for (int line = 0; line < linesLength; line++) {
+				try {
+					model.lines.addLine(line);
+				} catch (Exception e) {
+					e.printStackTrace();
+				}
+			}
+			model.lines.forEach(ModelLine::resetTokenizationState);
+			model.lines.get(0).setState(tokenizer.getInitialState());
+			model._invalidLineStartIndex = 0;
+			
+			while (!interrupted()) {

```

Comment:
With my Windows OS, even if thread is interrupted (when editor is closed), the while bloc content is executed and my CPU is growing.